### PR TITLE
Input: Batched reads & batched table allocations

### PR DESCRIPTION
**Pull request by NiLuJe**
Okay, turns out, reading the full batch of events at once (or at least in as few `read` as possible) isn't too gnarly, and allows us to do nifty things with Lua table allocations, by telling Lua exactly how many elements we're going to insert.

---
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader-base/1346)

### CODE REVIEW

**Comment by NiLuJe** [comment]
(I'll punt the debug printfs after dinner ;)).

File: input/input.c [hunk 1]
```diff
@@ -317,28 +344,70 @@ static int waitForInput(lua_State* L)
...
+                        // Kernel queue drained :)
+                        break;
+                    }
+                    lua_pop(L, lua_gettop(L));  // Kick our bogus bool (and potentially the ev_array table) from the stack
+                    lua_pushboolean(L, false);
+                    lua_pushinteger(L, errno);
+                    return 2;  // false, errno
+                }
+                if (len == 0) {
+                    // Should never happen
```
Comment (Frenzie) [inline input/input.c:372]: lol

File: ffi/posix_h.lua [hunk 2]
```diff
@@ -160,5 +160,7 @@ int sched_yield(void) __attribute__((nothrow, leaf));
 -- clock_gettime & friends require librt on old glibc (< 2.17) versions...
 if ffi.os == "Linux" then
     -- Load it in the global namespace to make it easier on callers...
-    pcall(ffi.load, "/lib/librt.so.1", true)
+    -- NOTE: There's no librt.so symlink, so, specify the SOVER, but not the full path,
+    --       in order to let the dynamic loader figure it out on its own.
+    pcall(ffi.load, "rt.so.1", true)
```
Comment (poire-z) [inline ffi/posix_h.lua:165]: (Just letting you know this keeps working on my GloHD.)

Comment (NiLuJe) [reply ffi/posix_h.lua:165]: Cool, thanks ;).

I'd checked with a dummy library and a strace that it was indeed following all the crazy stock paths, but glad to have a confirmation ;).

**Comment by NiLuJe** [comment]
Wheeee \o/

```
[pid 25222] 23:24:57.828218 [2ac7ae90] _newselect(8, [5</dev/input/event0> 6</dev/input/event1> 7<pipe:[72753]>], NULL, NULL, {tv_sec=1697, tv_usec=236071}) = 1 (in [6], left {tv_sec=1692, tv_usec=250573}) <4.985557>
[pid 25222] 23:25:02.814233 [2ac72e2c] read(6</dev/input/event1>, "\256\213g`\211k\f\0\3\09\0\1\0\0\0\256\213g`\217k\f\0\3\0000\0\1\0\0\0"..., 4096) = 112 <0.000036>
[pid 25222] 23:25:02.814570 [2ac72e2c] read(6</dev/input/event1>, 0x7ec627d8, 3984) = -1 EAGAIN (Resource temporarily unavailable) <0.000028>
[pid 25222] 23:25:02.815060 [2ac828dc] timerfd_create(CLOCK_REALTIME, TFD_CLOEXEC|TFD_NONBLOCK) = 11<anon_inode:[timerfd]> <0.000049>
[pid 25222] 23:25:02.815320 [2ac828fc] timerfd_settime(11<anon_inode:[timerfd]>, TFD_TIMER_ABSTIME, {it_interval={tv_sec=0, tv_nsec=0}, it_value={tv_sec=1617398703, tv_nsec=313977000}}, NULL) = 0 <0.000034>
[pid 25222] 23:25:02.816864 [2ac9031c] clock_gettime(CLOCK_MONOTONIC, {tv_sec=611936, tv_nsec=507595863}) = 0 <0.000034>
[pid 25222] 23:25:02.817232 [2ac7ae90] _newselect(12, [5</dev/input/event0> 6</dev/input/event1> 7<pipe:[72753]> 11<anon_inode:[timerfd]>], NULL, NULL, {tv_sec=1692, tv_usec=247057}) = 1 (in [6], left {tv_sec=1692, tv_usec=239888}) <0.007219>
[pid 25222] 23:25:02.824909 [2ac72e2c] read(6</dev/input/event1>, "\256\213g`\\\225\f\0\3\09\0\1\0\0\0\256\213g`b\225\f\0\3\0000\0\0\0\0\0"..., 4096) = 112 <0.000037>
[pid 25222] 23:25:02.825231 [2ac72e2c] read(6</dev/input/event1>, 0x7ec627d8, 3984) = -1 EAGAIN (Resource temporarily unavailable) <0.000026>
[pid 25222] 23:25:02.828375 [2ac9031c] clock_gettime(CLOCK_MONOTONIC, {tv_sec=611936, tv_nsec=519117922}) = 0 <0.000033>
[pid 25222] 23:25:02.828884 [2ac7ae90] _newselect(8, [5</dev/input/event0> 6</dev/input/event1> 7<pipe:[72753]>], NULL, NULL, {tv_sec=1692, tv_usec=235535}
```

**Review by NiLuJe — COMMENTED** [review]
> Reviewed 1 of 3 files at r1, 3 of 3 files at r3.
> Dismissed @poire-z from a discussion.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader-base/1346)* status: all files reviewed, 1 unresolved discussion (waiting on @NiLuJe)

___
*[ffi/posix_h.lua, line 165 at r2](https://reviewable.io/reviews/koreader/koreader-base/1346#-MXItqiX1PsfMz0B7aaC-r2-165:-MXJhrAS8N1cm8jwnnJq:b-896fix) ([raw file](https://github.com/koreader/koreader-base/blob/137a34383ee6aaa49f663051ca1fbed8cc22e79f/ffi/posix_h.lua#L165)):*
<details><summary><i>Previously, NiLuJe (NiLuJe) wrote…</i></summary><blockquote>

Cool, thanks ;).

I'd checked with a dummy library and a strace that it was indeed following all the crazy stock paths, but glad to have a confirmation ;).
</blockquote></details>

Done.

**Pull request merged**